### PR TITLE
Guiced up DSB dependency injection

### PR DESCRIPTION
**Pull request by DenverM80**
Handle injection binding in GuiModule

Implement providers for the various 'stores' that persist properties.

Also read the build version and date from a file generated at JAR build time.

Pull startup logic out of Main and into DeepStorageBrowser class

Implement ShutdownService

Inject or pass in ResourceBundle wherever encountered instances of re-opening.

Remove the "BackgroundTask" which consumed a thread to ping the BP every 5 seconds.

Clean up the CreateService by pulling out GetDataPolicies into its own task.

### CODE REVIEW

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/ShutdownServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                            jobId = ds3GetJob.getJobId().toString();
+                            ds3Client = ds3GetJob.getDs3Client();
+                        }
+                        LOG.info("Cancelled job:{} ", ds3GetJob.getJobId());
+                    } else if (job instanceof RecoverInterruptedJob) {
+                        final RecoverInterruptedJob recoverInterruptedJob = (RecoverInterruptedJob) job;
+                        jobId = recoverInterruptedJob.getUuid().toString();
+                        ds3Client = recoverInterruptedJob.getDs3Client();
+                        LOG.info("Cancelled job:{} ", recoverInterruptedJob.getUuid());
+                    }
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/ShutdownServiceImpl.java]: This is missing an else case.  Also it looks like the getDs3Client() call can be done first before this if else chain is called.  If it cannot, we should look to see what we can do to make it work that way.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/deletefiles/DeleteItemPresenter.java [hunk 2]
```diff
@@ -134,27 +140,22 @@ private void closeDialog() {
...
-            message = alertMessage + StringConstants.SPACE
-                    + ((Ds3DeleteFolderTask) deleteTask).getErrorMsg();
         } else if (deleteTask instanceof Ds3DeleteFilesTask) {
             alertMessage = resourceBundle.getString("deleteFailedError");
-            message = alertMessage + StringConstants.SPACE
-                    + ((Ds3DeleteFilesTask) deleteTask).getErrorMsg();
         }
+        message = alertMessage + StringConstants.SPACE + deleteTask.getErrorMsg();
 
+        LOG.error("Failed to delete selected item(s) ", message);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/deletefiles/DeleteItemPresenter.java]: This error message looks incorrect.  did you mean to add `{}` to the end of the string?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/Ds3PanelPresenter.java [hunk 3]
```diff
@@ -96,49 +98,49 @@
     @FXML
     private Ds3TreeTablePresenter ds3TreeTablePresenter;
 
-    @Inject
-    private Ds3SessionStore store;
-
-    @Inject
+    private ResourceBundle resourceBundle;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/Ds3PanelPresenter.java]: Some of these member variables look like they are missing final.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java [hunk 4]
```diff
@@ -66,40 +69,21 @@
...
-    private JobWorkers jobWorkers;
+    @ModelContext
+    private Ds3PanelPresenter ds3PanelPresenter;
 
-    @Inject
+    @ModelContext
     private Session session;
 
-    @Inject
+    private Workers workers;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java]: Missing final on a bunch of these member variables.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java [hunk 5]
```diff
@@ -198,6 +207,9 @@ private void initTreeTableView() {
 
         ds3TreeTable.expandedItemCountProperty().addListener((observable, oldValue, newValue) -> {
             if (ds3Common.getCurrentSession() != null) {
+                LOG.info("Loading Session " + session.getSessionName(), LogType.INFO);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java]: This log statement doe not look correct.  I assume it should be: 
```java
LOG.info("Loading Session {}", session.getSessionName());
```

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java [hunk 6]
```diff
@@ -50,32 +50,36 @@
     @FXML
     private TreeTableColumn jobIdColumn;
 
-    @Inject
+    private ResourceBundle resourceBundle;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: Missing final on several member variables.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/localfiletreetable/LocalFileTreeTablePresenter.java [hunk 7]
```diff
@@ -72,50 +74,54 @@
...
     private JobWorkers jobWorkers;
-
-    @Inject
-    private Ds3SessionStore store;
-
-    @Inject
-    private ResourceBundle resourceBundle;
-
-    @Inject
+    private Ds3SessionStore ds3SessionStore;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/localfiletreetable/LocalFileTreeTablePresenter.java]: Same comment about missing finals

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/metadata/MetadataPresenter.java [hunk 8]
```diff
@@ -18,33 +19,43 @@
...
+
     @FXML
     private TableColumn<MetadataEntry, String> metadataTableColValue;
 
     @FXML
     private TableColumn metadataTableColName;
 
-    @Inject
     private Ds3Metadata ds3Metadata;
+    private ResourceBundle resourceBundle;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/metadata/MetadataPresenter.java]: finals

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/modifyjobpriority/ModifyJobPriorityPresenter.java [hunk 9]
```diff
@@ -34,20 +36,24 @@
     @FXML
     private Button yesButton, noButton;
 
-    @Inject
     private ResourceBundle resourceBundle;
+    private Workers workers;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/modifyjobpriority/ModifyJobPriorityPresenter.java]: finals

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/BuildInfoServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(BuildInfoServiceImpl.class);
+
+    private static final String buildPropertiesFile = "dsb_build.properties";
+    private static final String buildPropertiesVersion = "version";
+    private static final String buildPropertiesDate = "build.date";
+
+    static final String dateTimeFormatterPattern = "EEE MMM dd kk:mm:ss z yyyy";  // build.date="Wed Mar 15 11:10:25 MDT 2017"
+    private static final DateTimeFormatter dtf = DateTimeFormatter.ofPattern(dateTimeFormatterPattern);
+
+    private String buildVersion;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/BuildInfoServiceImpl.java]: missing finals

File: dsb-gui/src/main/resources/dsb_build.properties [hunk 11]
```diff
@@ -0,0 +1,2 @@
+version=debug
```
Comment (rpmoore) [inline dsb-gui/src/main/resources/dsb_build.properties]: I don't think this file was meant to be checked in.

**Comment by DenverM80** [comment]
Updated, all requested changes fixed.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/ShutdownServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+        @Override
+        public Object call() {
+            outstandingJobs.forEach(job -> {
+                final CountDownLatch latch = new CountDownLatch(1);
+                try {
+
+                    String jobId = "";
+                    final Ds3Client ds3Client = job.getDs3Client();
+                    if (job instanceof Ds3PutJob) {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/ShutdownServiceImpl.java]: Did we discuss Adding the `getJobId()` method to the parent class of Ds3PutJob?  If we did that this whole code segment would collapse down into a single block.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/deletefiles/DeleteItemPresenter.java [hunk 13]
```diff
@@ -80,7 +87,7 @@ public void initialize(final URL location, final ResourceBundle resources) {
             });
 
         } catch (final Exception e) {
-            LOG.error("Encountered an error making the delete file presenter", e);
+            LOG.error("Encountered an error making the delete file presenter. \n{}", e);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/deletefiles/DeleteItemPresenter.java]: This is not the correct formatting for an error message.  If you want the message, then you need to use string concatenation.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/deletefiles/DeleteItemPresenter.java [hunk 14]
```diff
@@ -110,12 +117,11 @@ private void changeLabelText(final ObservableList<TreeItem<Ds3TreeTableValue>> s
     }
 
     public void deleteItems() {
-
         deleteTask.setOnCancelled(event -> constructMessageForLog());
         deleteTask.setOnFailed(event -> constructMessageForLog());
         deleteTask.setOnSucceeded(event -> {
             loggingService.logMessage(resourceBundle.getString("deleteSuccess"), LogType.SUCCESS);
-            LOG.info("Success to delete selected item(s):{} ", LogType.SUCCESS);
+            LOG.info("Successfully deleted selected item(s):{} ", LogType.SUCCESS);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/deletefiles/DeleteItemPresenter.java]: I know you didn't write this log message, but the `{}` doesn't tell us anything useful.  Remove that and the `LogType.SUCCESS`

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/metadata/MetadataPresenter.java [hunk 15]
```diff
@@ -18,33 +20,44 @@
...
+
     @FXML
     private TableColumn<MetadataEntry, String> metadataTableColValue;
 
     @FXML
     private TableColumn metadataTableColName;
 
-    @Inject
-    private Ds3Metadata ds3Metadata;
+    @ModelContext
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/metadata/MetadataPresenter.java]: Is this injected from the view?  If so it should not be passed in the constructor, and should be made non-final.

**Comment by DenverM80** [comment]
updated once more

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3JobTask.java [hunk 16]
```diff
@@ -43,7 +43,7 @@ protected final Boolean call() throws Exception {
         try {
             executeJob();
         } catch (final Exception e) {
-            LOG.error("Job failed with an exception", e);
+            LOG.error("Job failed with an exception: "+ e);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3JobTask.java]: This needs to be updated to:

```java

LOG.error("Job failed with an exception: " + e.getMessage() , e);
```

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3JobTask.java [hunk 17]
```diff
@@ -73,20 +75,21 @@ AtomicLong addDataTransferListener(final long totalJobSize) {
...
-     void addWaitingForChunkListener(final long totalJobSize, final String targetDir) {
+    void addWaitingForChunkListener(final long totalJobSize, final String targetDir) {
         job.attachWaitingForChunksListener(retryAfterSeconds -> {
             for (int retryTimeRemaining = retryAfterSeconds; retryTimeRemaining >= 0; retryTimeRemaining--) {
                 try {
                     updateMessage(resourceBundle.getString("noAvailableChunks") + SPACE + retryTimeRemaining + resourceBundle.getString("seconds"));
                     Thread.sleep(1000);
                 } catch (final Exception e) {
-                    LOG.error("Exception in attachWaitingForChunksListener: {}", e);
+                    LOG.error("Failed : ", e);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3JobTask.java]: This error message should say what failed, ie, what were we trying to do when it failed.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3PutJob.java [hunk 18]
```diff
@@ -146,19 +158,19 @@ public void executeJob() throws Exception {
             }
         } catch (final RuntimeException rte) {
             //cancel the job if it is already running
-            LOG.error("Encountered an error on a put job", rte);
             isJobFailed = true;
+            LOG.error("Encountered an error on a put job:\n", rte);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3PutJob.java]: Remove the newline character

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3PutJob.java [hunk 19]
```diff
@@ -146,19 +158,19 @@ public void executeJob() throws Exception {
...
-            LOG.error("Encountered an error on a put job: {}", ie);
-            loggingService.logMessage(StringBuilderUtil.jobCancelled(JobRequestType.PUT.toString()).toString()
-                    , LogType.ERROR);
+            LOG.error("Encountered an error on a put job:\n", ie);
+
+            loggingService.logMessage(StringBuilderUtil.jobCancelled(JobRequestType.PUT.toString()).toString(), LogType.ERROR);
         } catch (final Exception e) {
             isJobFailed = true;
-            LOG.error("Encountered an error on a put job: {}", e);
+            LOG.error("Encountered an error on a put job:\n", e);
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3PutJob.java]: Remove the new line character

**Pull request merged**